### PR TITLE
Adds a k8s deployment for the app, and updates Sonarcloud

### PR DESCRIPTION
**Pull request by ElvenSpellmaker**
#### 📲 What

Adds Infra and App deployment to the pipeline.

Also re-works Sonarscanner to work with Branches and PRs. (might need further tweaking due to `refs/heads`

#### 🤔 Why

We need our infra deployed and proper branch analysis on Sonarcloud.

#### 👀 Evidence

image
Now supports branches

#### 🕵️ How to test

Notes for QA

#### ✅ Acceptance criteria Checklist

- [ ] Code peer reviewed?
- [ ] Documentation has been updated to reflect the changes?
- [x] Passing all automated tests, including a successful deployment?
- [ ] Passing any exploratory testing?
- [x] Rebased/merged with latest changes from development and re-tested?
- [x] Meeting the Coding Standards?

### CODE REVIEW

File: build/azDevops/azure/templates/steps/build-java.yml [hunk 1]
```diff
@@ -146,15 +167,56 @@ steps:
 
   # SonarCloud
   - ${{ if eq(parameters.static_code_analysis, true) }}:
+      - bash: |
+          set -euxo pipefail
+          function strip_refs()
+          {
+            local BRANCH_REF="$1"
+
+            RETURN_BRANCH="$(sed -e "s%^refs/\(heads\|tags\)/%%" <<< $BRANCH_REF)"
```
Comment (dnitsch) [inline build/azDevops/azure/templates/steps/build-java.yml:176]: why don't you just echo this out as opposed to making a global variable?

File: build/azDevops/azure/templates/steps/build-java.yml [hunk 2]
```diff
@@ -146,15 +167,56 @@ steps:
...
+            EXTRA_SONAR_ARGUMENTS="${EXTRA_SONAR_ARGUMENTS} -Dsonar.branch.target='$RETURN_BRANCH'"
+          else
+            PROVIDER_LOWERCASED="$(tr '[:upper:]' '[:lower:]' <<< "${{ parameters.sonar_pullrequest_provider }}")"
+
+            EXTRA_SONAR_ARGUMENTS="-Dsonar.pullrequest.key='${{ parameters.pullrequest_number }}'"
+            strip_refs "${{ parameters.source_branch_ref }}"
+            EXTRA_SONAR_ARGUMENTS="${EXTRA_SONAR_ARGUMENTS} -Dsonar.pullrequest.branch='$RETURN_BRANCH'"
+            EXTRA_SONAR_ARGUMENTS="${EXTRA_SONAR_ARGUMENTS} -Dsonar.pullrequest.provider='$PROVIDER_LOWERCASED'"
+            strip_refs "${{ parameters.target_branch_ref }}"
+            EXTRA_SONAR_ARGUMENTS="${EXTRA_SONAR_ARGUMENTS} -Dsonar.pullrequest.base'$RETURN_BRANCH'"
```
Comment (dnitsch) [inline build/azDevops/azure/templates/steps/build-java.yml:192]: is the intention here to just overwrite the `EXTRA_SONAR_ARGUMENTS` ? or am I misreading this?

Comment (ElvenSpellmaker) [reply build/azDevops/azure/templates/steps/build-java.yml:192]: `EXTRA_SONAR_ARGUMENTS="${EXTRA_SONAR_ARGUMENTS} ...` appends something to the end of the var

File: build/azDevops/azure/templates/steps/build-java.yml (hunk 1 above)
Comment (ElvenSpellmaker) [reply build/azDevops/azure/templates/steps/build-java.yml:176]: Could do, but echoing is prone to failures, if the method did anything else. Plus we have `-x` set which means we'd get the command running output in the output of the function iirc.

File: build/azDevops/azure/templates/steps/build-java.yml (hunk 2 above)
Comment (ElvenSpellmaker) [reply build/azDevops/azure/templates/steps/build-java.yml:192]: Although I see I'm missing an `=` near the end of that line.

File: build/azDevops/azure/templates/steps/build-java.yml [hunk 3]
```diff
@@ -146,15 +167,56 @@ steps:
...
+          fi
+
+          echo "##vso[task.setvariable variable=bash_extra_sonar_arguments]$EXTRA_SONAR_ARGUMENTS"
+        displayName: "Sonar: Extra Sonar Arguments"
+
       - template: azDevOps/azure/templates/v2/steps/test-static-code-sonar.yml@templates
         parameters:
           working_directory: "${{ parameters.project_root_dir }}"
           target_container: "${{ parameters.sonar_container }}"
+          extra_arguments: "$(bash_extra_sonar_arguments)"
```
Comment (dnitsch) [inline build/azDevops/azure/templates/steps/build-java.yml:206]: why not make this EXTRA_SONAR_ARGS env_vars and default to ""? 
e.g. 
```env_vars: { 
   ... 
  EXTRA_SONAR_ARGS: "-Dsonar.pullrequest.github.repository='${{ parameters.sonar_remote_repo }}' -Dsonar.pullrequest.key='${{ parameters.pullrequest_number }}'  -Dsonar.pullrequest.branch='$RETURN_BRANCH' -Dsonar.pullrequest.provider='$PROVIDER_LOWERCASED'"
```

Comment (dnitsch) [reply build/azDevops/azure/templates/steps/build-java.yml:212]: I presume inside the function you are appending it after `sonar-scanner ${{ parameters.extra_arguments }}`  and defaulted the parameter to `""` instead you could also try this 
```
sonar-scanner ${EXTRA_SONAR_ARGS=""}
```

Comment (ElvenSpellmaker) [reply build/azDevops/azure/templates/steps/build-java.yml:212]: Hmmm how so, does that work?

Comment (dnitsch) [reply build/azDevops/azure/templates/steps/build-java.yml:212]: that's a bash fallback value to "" in cases where users don't specify it. you could actually use ${EXTRA_SONAR_ARGS:=""}

Comment (ElvenSpellmaker) [reply build/azDevops/azure/templates/steps/build-java.yml:212]: I can change it, but is there a reason to do it that way? I was thinking this way it's clear the arguments are appended to the command.

I don't mind either way.

Comment (ElvenSpellmaker) [reply build/azDevops/azure/templates/steps/build-java.yml:212]: The other way you mentioned would pass it as an env var as well as extra arguments which might be unintended?

**Pull request merged**